Import Shape Detection API IDLs

litherum opened the pull request:
#### c1ae4bf56dcfbb654055cb817d22eca9f6f0dd98
<pre>
Import Shape Detection API IDLs
<a href="https://bugs.webkit.org/show_bug.cgi?id=255191">https://bugs.webkit.org/show_bug.cgi?id=255191</a>
rdar://107783367

Reviewed by Alex Christensen.

This creates a new setting, named ShapeDetection, that is off by default, and
imports the Shape Detection API IDLs guarded by that setting. The IDLs are
current as of cc6c9460a134e53901da65822dfd36cc11b4331b. The methods backing
the IDLs are just stubbed out and do nothing just yet.

See:
<a href="https://github.com/WICG/shape-detection-api">https://github.com/WICG/shape-detection-api</a>
<a href="https://wicg.github.io/shape-detection-api/">https://wicg.github.io/shape-detection-api/</a>
<a href="https://wicg.github.io/shape-detection-api/text.html">https://wicg.github.io/shape-detection-api/text.html</a>

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.cpp: Added.
(WebCore::BarcodeDetector::getSupportedFormats):
(WebCore::BarcodeDetector::detect):
* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.h: Added.
(WebCore::BarcodeDetector::create):
(WebCore::BarcodeDetector::BarcodeDetector):
* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.idl: Added.
* Source/WebCore/Modules/ShapeDetection/BarcodeDetectorOptions.h: Added.
* Source/WebCore/Modules/ShapeDetection/BarcodeDetectorOptions.idl: Added.
* Source/WebCore/Modules/ShapeDetection/BarcodeFormat.h: Added.
* Source/WebCore/Modules/ShapeDetection/BarcodeFormat.idl: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedBarcode.h: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedBarcode.idl: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedFace.h: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedFace.idl: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedText.h: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedText.idl: Added.
* Source/WebCore/Modules/ShapeDetection/FaceDetector.cpp: Added.
(WebCore::FaceDetector::detect):
* Source/WebCore/Modules/ShapeDetection/FaceDetector.h: Added.
(WebCore::FaceDetector::create):
(WebCore::FaceDetector::FaceDetector):
* Source/WebCore/Modules/ShapeDetection/FaceDetector.idl: Added.
* Source/WebCore/Modules/ShapeDetection/FaceDetectorOptions.h: Added.
* Source/WebCore/Modules/ShapeDetection/FaceDetectorOptions.idl: Added.
* Source/WebCore/Modules/ShapeDetection/Landmark.h: Added.
* Source/WebCore/Modules/ShapeDetection/Landmark.idl: Added.
* Source/WebCore/Modules/ShapeDetection/LandmarkType.h: Added.
* Source/WebCore/Modules/ShapeDetection/LandmarkType.idl: Added.
* Source/WebCore/Modules/ShapeDetection/Point2D.h: Added.
* Source/WebCore/Modules/ShapeDetection/Point2D.idl: Added.
* Source/WebCore/Modules/ShapeDetection/TextDetector.cpp: Added.
(WebCore::TextDetector::detect):
* Source/WebCore/Modules/ShapeDetection/TextDetector.h: Added.
(WebCore::TextDetector::create):
(WebCore::TextDetector::TextDetector):
* Source/WebCore/Modules/ShapeDetection/TextDetector.idl: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

Canonical link: <a href="https://commits.webkit.org/262755@main">https://commits.webkit.org/262755@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/303c11c2c61324614d08754b49f7dfc909d61f64

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2520 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2522 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2645 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3920 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2554 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2485 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2663 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2609 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/2226 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2543 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/2596 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2263 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3684 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/368 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2248 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/2104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/2100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2474 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2268 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/3449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/2414 "Built successfully and passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2298 "Passed tests") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2081 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2596 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2295 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2251 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/598 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/617 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2273 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2651 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2424 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/3/builds/711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/Modules/ShapeDetection/BarcodeDetector.idl:
```diff
@@ -0,0 +1,56 @@
+
+// https://wicg.github.io/shape-detection-api/#barcodedetector
+
+// FIXME: This shouldn't need to be duplicated here.
```
Bug number?

litherum commented on Source/WebCore/Modules/ShapeDetection/BarcodeDetector.idl:
```diff
@@ -0,0 +1,56 @@
+
+[
+    EnabledBySetting=ShapeDetection,
+    Exposed=(Window), /* https://bugs.webkit.org/show_bug.cgi?id=232542: Worker */
```
Is this the right bug?

achristensen07 commented on Source/WebCore/Modules/ShapeDetection/DetectedBarcode.h:
```diff
@@ -0,0 +1,44 @@
+namespace WebCore {
+
+struct DetectedBarcode {
+    RefPtr<DOMRectReadOnly> boundingBox;
```
Can this be a Ref?  IDL says "required"

litherum replied:
Unfortunately, it can't :( our bindings code will try to default-initialize these structs :(

achristensen07 commented on Source/WebCore/Modules/ShapeDetection/DetectedFace.h:
```diff
@@ -0,0 +1,41 @@
+namespace WebCore {
+
+struct DetectedFace {
+    RefPtr<DOMRectReadOnly> boundingBox;
```
ditto, several other places.

achristensen07 commented on Source/WebCore/Modules/ShapeDetection/DetectedFace.h:
```diff
@@ -0,0 +1,41 @@
+
+#pragma once
+
+#include "DOMRectReadOnly.h"
```
If we put constructor and destructor in a C++ file, this and Landmark.h can be just forward declarations.  Saves compile time.

achristensen07 commented on Source/WebCore/Modules/ShapeDetection/FaceDetectorOptions.h:
```diff
@@ -0,0 +1,35 @@
+namespace WebCore {
+
+struct FaceDetectorOptions {
+    unsigned short maxDetectedFaces;
```
uint16_t?
default initialize to avoid uninitialized memory.
I think this is a strange type, but that's what the spec says right now.

achristensen07 commented on Source/WebCore/Modules/ShapeDetection/Landmark.h:
```diff
@@ -0,0 +1,39 @@
+
+struct Landmark {
+    Vector<Point2D> locations;
+    LandmarkType type;
```
default initialize?

achristensen07 commented on Source/WebCore/Modules/ShapeDetection/Point2D.h:
```diff
@@ -0,0 +1,35 @@
+namespace WebCore {
+
+struct Point2D {
+    double x = 0;
```
double x { 0 }

litherum committed:
Import Shape Detection API IDLs
https://bugs.webkit.org/show_bug.cgi?id=255191
rdar://107783367

Reviewed by Alex Christensen.

This creates a new setting, named ShapeDetection, that is off by default, and
imports the Shape Detection API IDLs guarded by that setting. The IDLs are
current as of cc6c9460a134e53901da65822dfd36cc11b4331b. The methods backing
the IDLs are just stubbed out and do nothing just yet.

See:
https://github.com/WICG/shape-detection-api
https://wicg.github.io/shape-detection-api/
https://wicg.github.io/shape-detection-api/text.html

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.cpp: Added.
(WebCore::BarcodeDetector::getSupportedFormats):
(WebCore::BarcodeDetector::detect):
* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.h: Added.
(WebCore::BarcodeDetector::create):
(WebCore::BarcodeDetector::BarcodeDetector):
* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.idl: Added.
* Source/WebCore/Modules/ShapeDetection/BarcodeDetectorOptions.h: Added.
* Source/WebCore/Modules/ShapeDetection/BarcodeDetectorOptions.idl: Added.
* Source/WebCore/Modules/ShapeDetection/BarcodeFormat.h: Added.
* Source/WebCore/Modules/ShapeDetection/BarcodeFormat.idl: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedBarcode.h: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedBarcode.idl: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedFace.h: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedFace.idl: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedText.h: Added.
* Source/WebCore/Modules/ShapeDetection/DetectedText.idl: Added.
* Source/WebCore/Modules/ShapeDetection/FaceDetector.cpp: Added.
(WebCore::FaceDetector::detect):
* Source/WebCore/Modules/ShapeDetection/FaceDetector.h: Added.
(WebCore::FaceDetector::create):
(WebCore::FaceDetector::FaceDetector):
* Source/WebCore/Modules/ShapeDetection/FaceDetector.idl: Added.
* Source/WebCore/Modules/ShapeDetection/FaceDetectorOptions.h: Added.
* Source/WebCore/Modules/ShapeDetection/FaceDetectorOptions.idl: Added.
* Source/WebCore/Modules/ShapeDetection/Landmark.h: Added.
* Source/WebCore/Modules/ShapeDetection/Landmark.idl: Added.
* Source/WebCore/Modules/ShapeDetection/LandmarkType.h: Added.
* Source/WebCore/Modules/ShapeDetection/LandmarkType.idl: Added.
* Source/WebCore/Modules/ShapeDetection/Point2D.h: Added.
* Source/WebCore/Modules/ShapeDetection/Point2D.idl: Added.
* Source/WebCore/Modules/ShapeDetection/TextDetector.cpp: Added.
(WebCore::TextDetector::detect):
* Source/WebCore/Modules/ShapeDetection/TextDetector.h: Added.
(WebCore::TextDetector::create):
(WebCore::TextDetector::TextDetector):
* Source/WebCore/Modules/ShapeDetection/TextDetector.idl: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

Canonical link: https://commits.webkit.org/262755@main
